swap/swap.go

```diff
@@ -17,69 +17,340 @@
+}
+
+// Add is the (sole) accounting function
+// Swap implements the protocols.Balance interface
 func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
-	//load existing balances from the state store
+	// load existing balances from the state store
 	err = s.loadState(peer)
 	if err != nil && err != state.ErrNotFound {
+		log.Error("error while loading balance for peer", "peer", peer.ID().String())
 		return
 	}
+
+	// Check if balance with peer is over the disconnect threshold
+	// It is the creditor who triggers the disconnect from a overdraft creditor,
+	// thus we check for a positive value
+	if s.balances[peer.ID()] >= s.disconnectThreshold {
+		// if so, return error in order to abort the transfer
+		disconnectMessage := fmt.Sprintf("balance for peer %s is over the disconnect threshold %v, disconnecting", peer.ID().String(), s.disconnectThreshold)
+		log.Warn(disconnectMessage)
+		return errors.New(disconnectMessage)
+	}
+
+	// calculate new balance
+	var newBalance int64
+	newBalance, err = s.updateBalance(peer.ID(), amount)
+	if err != nil {
+		return
+	}
+
+	// Check if balance with peer crosses the threshold
+	// It is the peer with a negative balance who sends a cheque, thus we check
+	// that the balance is *below* the threshold
+	if newBalance <= -s.paymentThreshold {
+		//if so, send cheque
+		log.Warn(fmt.Sprintf("balance for peer %s went over the payment threshold %v, sending cheque", peer.ID().String(), s.paymentThreshold))
+		err = s.sendCheque(peer.ID())
+		if err != nil {
+			log.Error(fmt.Sprintf("error while sending cheque to peer %s: %s", peer.ID().String(), err.Error()))
+		} else {
+			log.Info(fmt.Sprintf("successfully sent cheque to peer %s", peer.ID().String()))
+		}
+	}
+
+	return
+}
+
+func (s *Swap) updateBalance(peer enode.ID, amount int64) (int64, error) {
 	//adjust the balance
 	//if amount is negative, it will decrease, otherwise increase
-	s.balances[peer.ID()] += amount
+	s.balances[peer] += amount
 	//save the new balance to the state store
-	peerBalance := s.balances[peer.ID()]
-	err = s.stateStore.Put(peer.ID().String(), &peerBalance)
+	peerBalance := s.balances[peer]
+	err := s.stateStore.Put(peer.String(), &peerBalance)
+	if err != nil {
+		log.Error("error while storing balance for peer", "peer", peer.String())
+	}
+	log.Debug("balance for peer after accounting", "peer", peer.String(), "balance", strconv.FormatInt(peerBalance, 10))
+	return peerBalance, err
+}
 
-	log.Debug(fmt.Sprintf("balance for peer %s: %s", peer.ID().String(), strconv.FormatInt(peerBalance, 10)))
+// logBalance is a helper function to log the current balance of a peer
+func (s *Swap) logBalance(peer *protocols.Peer) {
+	err := s.loadState(peer)
+	if err != nil && err != state.ErrNotFound {
+		log.Error(fmt.Sprintf("error while loading balance for peer %s", peer.String()))
```

Comment:
return err. don't check for the err not found - this is also easy to assert from the calling context. keep it simple:
```
if err != nil {
 return err
}
....

Reply:
Not sure I follow here. If `err != nil` but `err == state.ErrNotFound`, we want to continue with the rest of the function.

Reply:
The logBalance function was removed due to a previous comment on this pull request from acud. As a result this is no longer a problem, so we can consider it fixed.